File: src/main/native/cpp/Rev2mDistanceSensor.cpp
#include "Rev2mDistanceSensor.h"
#include "vl53l0x_api.h"

#include <hal/HAL.h>
#include <hal/I2C.h>

#include <thread>
#include <atomic>
#include <vector>
#include <memory>

#include <frc/Timer.h>
#include <frc/DriverStation.h>
#include <frc/smartdashboard/SendableBase.h>
#include <frc/smartdashboard/SendableBuilder.h>
#include <frc/Utility.h>

#define USE_I2C_2V8
//#define _DEBUG_

using namespace rev;

std::atomic<bool> Rev2mDistanceSensor::m_automaticEnabled{false};
std::atomic<double> Rev2mDistanceSensor::m_measurementPeriod{0.05};
std::vector<Rev2mDistanceSensor*> Rev2mDistanceSensor::m_sensors;
std::thread Rev2mDistanceSensor::m_thread;

void print_pal_error(VL53L0X_Error Status){
    char buf[VL53L0X_MAX_STRING_LENGTH];
    VL53L0X_GetPalErrorString(Status, buf);
    printf("API Status: %i : %s\n", Status, buf);
}

Rev2mDistanceSensor::Rev2mDistanceSensor(Port port, int deviceAddress, DistanceUnit units)
    : m_port(static_cast<HAL_I2CPort>(port)), m_deviceAddress(deviceAddress) {

    pDevice->I2cDevAddr = deviceAddress;
    pDevice->port = m_port;
    m_units = units;

    int32_t status = 0;
    HAL_InitializeI2C(m_port, &status);

    HAL_Report(HALUsageReporting::kResourceType_I2C, deviceAddress);

    if(!Initialize())
        frc::DriverStation::ReportWarning("Error initializing Rev 2M device.");
}

Rev2mDistanceSensor::~Rev2mDistanceSensor() { 
    bool wasAutomaticMode = m_automaticEnabled;
    SetAutomaticMode(false);

    // No synchronization needed because the background task is stopped.
    m_sensors.erase(std::remove(m_sensors.begin(), m_sensors.end(), this),
                    m_sensors.end());

    if (!m_sensors.empty() && wasAutomaticMode) {
        SetAutomaticMode(true);
    }    
}

bool Rev2mDistanceSensor::IsRangeValid() const {
    return m_rangeValid;
}

double Rev2mDistanceSensor::GetRange(DistanceUnit units) {
    if(units == kCurrent)
        units = m_units;

    switch (units) {
        case Rev2mDistanceSensor::kInches:
            return GetRangeInches();
        case Rev2mDistanceSensor::kMilliMeters:
            return GetRangeMM();
        default:
            return -1;
    }
}

double Rev2mDistanceSensor::GetTimestamp(void) {
    return m_timestamp;
}

double Rev2mDistanceSensor::GetRangeMM() {
    return m_currentRange;
}

double Rev2mDistanceSensor::GetRangeInches() {
    return m_currentRange / 25.4;
}

void Rev2mDistanceSensor::SetEnabled(bool enable) { m_enabled = enable; }

void Rev2mDistanceSensor::SetRangeProfile(RangeProfile profile) {
    uint32_t budget;

    if (profile == RangeProfile::kHighAccuracy) {
        m_profile = RangeProfile::kHighAccuracy;
        budget = 200000;
    }
    else if (profile == RangeProfile::kLongRange) {
        m_profile = RangeProfile::kLongRange;
        budget = 33000;
    }
    else if (profile == RangeProfile::kHighSpeed) {
        m_profile = RangeProfile::kLongRange;
        budget = 20000;
    }
    else {
        m_profile = RangeProfile::kLongRange;
        budget = 20000;
    }

    if(m_stopped) {
        #ifdef _DEBUG_
        printf("Sensor stopped. Setting timing budget\n");
        #endif
        SetMeasurementTimingBudget(budget);
    }
    else {
        #ifdef _DEBUG_
        printf("Sensor disabled. Setting budget to %d\n", budget);
        #endif
        m_enabled = false;
        m_newMeasurementTimingBudget = budget;
    }
}

void Rev2mDistanceSensor::SetMeasurementTimingBudget(uint32_t budget_us) {
    Status = VL53L0X_SetMeasurementTimingBudgetMicroSeconds(pDevice, budget_us);

    if(Status == VL53L0X_ERROR_NONE) {
        m_currentMeasurementTimingBudget = budget_us;
        m_newMeasurementTimingBudget = budget_us;
        #ifdef _DEBUG_
        printf("Timing budget changed to %dus\n", m_currentMeasurementTimingBudget);
        #endif
    }
    else {
        #ifdef _DEBUG_
        printf("Error setting timing buget\n");
        #endif
    }
}

bool Rev2mDistanceSensor::GetMeasurementTimingBudget(uint32_t *budget_us) {
    Status = VL53L0X_GetMeasurementTimingBudgetMicroSeconds(pDevice, budget_us);

    return Status == 0;
}

void Rev2mDistanceSensor::SetMeasurementPeriod(double period) {
    if(period < 0.01) {
        period = 0.01;
    } 
    else if(period > 1) {
        period = 1;
    }
    
    m_measurementPeriod = period;
}

double Rev2mDistanceSensor::GetMeasurementPeriod(void) {
    return m_measurementPeriod;
}

void Rev2mDistanceSensor::SetDistanceUnits(DistanceUnit units) { m_units = units; }

Rev2mDistanceSensor::DistanceUnit Rev2mDistanceSensor::GetDistanceUnits() const {
  return m_units;
}

bool Rev2mDistanceSensor::Initialize(void) {
    uint32_t refSpadCount;
    uint8_t isApertureSpads;
    uint8_t VhvSettings;
    uint8_t PhaseCal;
    
    #ifdef  _DEBUG_
    printf("Initializing device on port: ");
    if(m_port == HAL_I2C_kOnboard)
        printf("Onboard\n");
    else
        printf("MXP\n");
    #endif

    if(ValidateI2C() != true)
        frc::DriverStation::ReportError("Error communicating with Rev 2M sensor over I2C.");

    VL53L0X_Version_t *pVersion = new VL53L0X_Version_t;
    Status = VL53L0X_GetVersion(pVersion);
    #ifdef  _DEBUG_
    if(Status != VL53L0X_ERROR_NONE) {
        frc::DriverStation::ReportError("Version error.");
        print_pal_error(Status);
    }
    #endif

    Status = VL53L0X_DataInit(pDevice); // Data initialization
    #ifdef _DEBUG_
    if(Status != VL53L0X_ERROR_NONE) {
        frc::DriverStation::ReportError("Data init error.");
        print_pal_error(Status);
    }
    #endif

    VL53L0X_DeviceInfo_t *pDeviceInfo = new VL53L0X_DeviceInfo_t;
    Status = VL53L0X_GetDeviceInfo(pDevice, pDeviceInfo);
    #ifdef _DEBUG_
    if(Status != VL53L0X_ERROR_NONE) {
        frc::DriverStation::ReportError("Get device info error.");
        print_pal_error(Status);
    }
    #endif
    
    Status = VL53L0X_StaticInit(pDevice); // Device Initialization
    #ifdef _DEBUG_
    if(Status != VL53L0X_ERROR_NONE) {
        frc::DriverStation::ReportError("Static init error.");
        print_pal_error(Status);
    }
    #endif
    
    Status = VL53L0X_PerformRefCalibration(pDevice, &VhvSettings, &PhaseCal); // Device Initialization
    #ifdef _DEBUG_
    if(Status != VL53L0X_ERROR_NONE) {
        frc::DriverStation::ReportError("Static init error.");
        print_pal_error(Status);
    }
    #endif

    Status = VL53L0X_PerformRefSpadManagement(pDevice, &refSpadCount, &isApertureSpads); // Device Initialization
    #ifdef _DEBUG_
    if(Status != VL53L0X_ERROR_NONE) {
        frc::DriverStation::ReportError("Erroring performing ref spad management.");
        print_pal_error(Status);
    }
    #endif

    Status = VL53L0X_SetDeviceMode(pDevice, VL53L0X_DEVICEMODE_CONTINUOUS_RANGING);
    #ifdef _DEBUG_
    if(Status != VL53L0X_ERROR_NONE) {
        frc::DriverStation::ReportWarning("Error starting Rev 2M sensor\n");
    }
    #endif

    m_sensors.emplace_back(this);
    m_enabled = true;

    // return true if initialization was successful
    return Status >= 0;
}

bool Rev2mDistanceSensor::ValidateI2C(void) {
    uint8_t reg = 0xC0, res[2];

    HAL_TransactionI2C(pDevice->port, pDevice->I2cDevAddr >> 1, &reg, 1, res, 1);

    if(*res != 0xEE)
        return false;

    reg = 0xC1;
    HAL_TransactionI2C(pDevice->port, pDevice->I2cDevAddr >> 1, &reg, 1, res, 1);

    if(*res != 0xAA)
        return false;

    reg = 0xC2;
    HAL_TransactionI2C(pDevice->port, pDevice->I2cDevAddr >> 1, &reg, 1, res, 1);

    if(*res != 0x10)
        return false;

    reg = 0x51;
    HAL_TransactionI2C(pDevice->port, pDevice->I2cDevAddr >> 1, &reg, 1, res, 2);

    reg = 0x61;
    HAL_TransactionI2C(pDevice->port, pDevice->I2cDevAddr >> 1, &reg, 1, res, 2);

    if((res[0] != 0x00) || (res[1] != 0x00))
        return false;

    return true;
}

bool Rev2mDistanceSensor::IsEnabled() const { return m_enabled; }

void Rev2mDistanceSensor::SetAutomaticMode(bool enabling)
{
    if (enabling == m_automaticEnabled) return;  // ignore the case of no change

    m_automaticEnabled = enabling;

    // if automatic is being enabled
    if(enabling) {
        // if thread is not currently running, start it
        if(!m_thread.joinable())
            #ifdef _DEBUG_
            printf("Starting thread\n");
            #endif
            m_thread = std::thread(&Rev2mDistanceSensor::DoContinuous);
    } 
    // else automatic mode is being disabled
    else {
        m_automaticEnabled = false;

        // disable all sensors
        for(auto& sensor : m_sensors)
            sensor->SetEnabled(false);
    }
}

void Rev2mDistanceSensor::DoContinuous(void) {
    VL53L0X_Error stat = VL53L0X_ERROR_NONE;
    VL53L0X_RangingMeasurementData_t *pRangingMeasurementData = new VL53L0X_RangingMeasurementData_t;
    uint8_t NewDatReady = 0;
    bool allStopped;

    #ifdef _DEBUG_
    printf("Thread started\n");
    #endif

    do {
        // used to stop loop. allStopped will be changed to false if any sensors still running
        allStopped = true;

        // iterate through sensors
        for(auto& sensor : m_sensors)
        {
            // if sensor is enabled
            if(sensor->IsEnabled()) {
                // if sensor has not been started yet
                if(sensor->m_stopped) {
                    // command device to start measurement
                    stat = VL53L0X_StartMeasurement(sensor->pDevice);

                    if(stat == VL53L0X_ERROR_NONE)
                        sensor->m_stopped = false;
                }
                // else sensor has been started
                else {
                    // check for new data
                    stat = VL53L0X_GetMeasurementDataReady(sensor->pDevice, 
                                                            &NewDatReady);

                    // if new data is ready
                    if((NewDatReady == 0x01) && (stat == VL53L0X_ERROR_NONE)) {
                        stat = VL53L0X_GetRangingMeasurementData(sensor->pDevice, 
                                    pRangingMeasurementData);
                        VL53L0X_ClearInterruptMask(sensor->pDevice, 
                                    VL53L0X_REG_SYSTEM_INTERRUPT_GPIO_NEW_SAMPLE_READY);

                        if(stat == VL53L0X_ERROR_NONE) {
                            // range is valid when RangeStatus equals 0
                            sensor->m_rangeValid = pRangingMeasurementData->RangeStatus == 0;
                            sensor->m_currentRange = pRangingMeasurementData->RangeMilliMeter;
                            sensor->m_timestamp = frc::Timer::GetFPGATimestamp();
                        } else {
                            sensor->m_rangeValid = false;
                        }
                    } else {
                        sensor->m_rangeValid = false;
                    }
                }
            } 
            // else automatic is not enabled
            else if(!sensor->m_stopped) {
                // if sensor is not currently stopping
                if(!sensor->m_stopping) {
                    // command device to stop measurements
                    stat = VL53L0X_StopMeasurement(sensor->pDevice);

                    if(stat == VL53L0X_ERROR_NONE)
                        sensor->m_stopping = true;
                }
                // else sensor is currently stopping
                else {
                    // check if sensor has finished
                    uint32_t StopCompleted = 0;
                    stat = VL53L0X_GetStopCompletedStatus(sensor->pDevice, &StopCompleted);

                    if ((StopCompleted == 0x00) || stat != VL53L0X_ERROR_NONE) {
                        sensor->m_stopped = true;
                        sensor->m_stopping = false;
                        stat = VL53L0X_ClearInterruptMask(sensor->pDevice, 
                                    VL53L0X_REG_SYSTEM_INTERRUPT_GPIO_NEW_SAMPLE_READY);
                    }
                }
            }
            // else if sensor is stopped and measurement timing budget needs to be changed
            else if(sensor->m_newMeasurementTimingBudget != sensor->m_currentMeasurementTimingBudget) {
                sensor->SetMeasurementTimingBudget(sensor->m_newMeasurementTimingBudget);

                if(sensor->m_newMeasurementTimingBudget == sensor->m_currentMeasurementTimingBudget) {
                    // if timing budget is successfully set, reenable sensor
                    sensor->m_enabled = true;
                }
            }

            // check if there are any sensors still enabled
            if(!sensor->m_stopped)
                allStopped = false;
        }

        frc::Wait(m_measurementPeriod);
    }
    while(!allStopped && m_automaticEnabled);

    #ifdef _DEBUG_
    printf("Stopping thread\n");
    #endif

    // all sensors stopped and automatic is disabled so detach thread
    m_thread.detach();
}

void Rev2mDistanceSensor::InitSendable(frc::SendableBuilder& builder) {
    builder.SetSmartDashboardType("Distance");
    builder.AddDoubleProperty("Value", [=]() { return GetRange(); }, nullptr);
}

double Rev2mDistanceSensor::PIDGet() {
    switch (m_units) {
        case Rev2mDistanceSensor::kInches:
            return GetRangeInches();
        case Rev2mDistanceSensor::kMilliMeters:
            return GetRangeMM();
        default:
            return 0.0;
    }
}

void Rev2mDistanceSensor::SetPIDSourceType(frc::PIDSourceType pidSource) {
    if (wpi_assert(pidSource == frc::PIDSourceType::kDisplacement)) {
        m_pidSource = pidSource;
    }
}

File: Source/src/main/java/com/revrobotics/Rev2mDistanceSensor.java
/*
 * Copyright (c) 2018-2020 REV Robotics
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of REV Robotics nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.revrobotics;

import java.util.ArrayList;
import java.util.List;

import com.revrobotics.jni.VL53L0XJNI;

import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.DriverStation;

/**
 * Rev2mDistanceSensor class.
 *
 * The Distance sensor class is intended to be used with the RevRobotics
 * 2M Distance Sensor.
 */
public class Rev2mDistanceSensor {
    public enum Port { kOnboard, kMXP }
    public enum RangeProfile { kDefault, kHighAccuracy, kLongRange, kHighSpeed }
    public enum Unit { kInches, kMillimeters }

    // sensor list
    private static final List<Rev2mDistanceSensor> m_sensors = new ArrayList<>();
    // automatic round robin mode
    private static volatile boolean m_automaticEnabled;
    private boolean m_enabled;
    // task doing the round-robin automatic sensing
    private static Thread m_task;
    private Unit m_units;

    private RangeProfile m_profile = RangeProfile.kDefault;
    private RangeProfile m_newProfile = RangeProfile.kDefault;
    private boolean m_stopped = true;
    private boolean m_rangeValid = false;
    private double m_currentRange = -1;
    private double m_timestamp = -1;
    private boolean m_stopping = false;
    private static double m_measurementPeriod = 0.05;
    private int m_addr = 0x53;
    private int m_port;

    /**
     * Creat an instance of the Rev 2m Distance Sensor.
     *
     * @param port          The I2C port to which the device is connected.
     * @param units         The units returned in either kInches or kMilliMeters
     * @param profile       Range profile for the device. Valid options are:
     *                          kDefault
     *                          kHighAccuracy
     *                          kLongRange
     *                          kHighSpeed
     *                      Refer to data sheet for profile specific performance specs.
     */
    public Rev2mDistanceSensor(Port port, Unit units, RangeProfile profile) {
        m_units = units;

        if(port == Port.kOnboard)
            m_port = 0;
        else
            m_port = 1;

        VL53L0XJNI.Init(m_port, m_addr);

        if(!initialize()) {
            DriverStation.reportError(String.format("Error initializing Rev 2M device on port " +
                                                    "%s. Please check your connections",
                                                    port == Port.kMXP ? "MXP" : "Onboard"), false);
        }

    }

    public Rev2mDistanceSensor(Port port) {
        this(port, Unit.kInches, RangeProfile.kDefault);
    }

    /**
     * Check if the device is enabled in round robin mode
     *
     * @return  True - Device enabled
     *          False - Device disabled
     */
    public boolean isEnabled() {
        return m_enabled;
    }

    /**
     * Turn automatic mode on/off.
     *
     * When automatic mode is on, all enabled sensors will be polled at a set rate.
     *
     * @param enabling  Set to true if round robin scheduling should start. This
     *                  will command all sensors to begin measurements in
     *                  continuous ranging mode.
     *
     *                  Setting to false will command all sensors to stop
     *                  measurements.
     */
    public void setAutomaticMode(boolean enabling) {
        if(enabling == m_automaticEnabled) return;

        m_automaticEnabled = enabling;

        if(enabling) {
            if(m_task == null) {
                m_task = new DoContinuous();
                m_task.start();
            }
            else if(m_task.isAlive()) {
                try {
                    m_task.join();
                    m_task = new DoContinuous();
                    m_task.start();
                } catch (InterruptedException ex) {
                    Thread.currentThread().interrupt();
                    ex.printStackTrace();
                }
            }
            else {
                m_task = new DoContinuous();
                m_task.start();
            }
        }
        else {
            for(Rev2mDistanceSensor sensor: m_sensors) {
                sensor.setEnabled(false);
            }
        }
    }

    /**
     * Determine if the current range is valid.
     *
     * @return  True - current range is valid
     *          False - current range is not valid
     */
    public boolean isRangeValid() {
        return m_rangeValid;
    }

    /**
     * Get the range in current units from the sensor.
     *
     * @return range in current units.
     */
    public double getRange() {
        return getRange(m_units);
    }

    /**
     * Get the range in specified units from the sensor.
     *
     * @param units units in which to return the measurement.
     *              Valid options are:
     *                  kInches
     *                  kMilliMeters
     *
     * @return range in specified units.
     */
    public double getRange(Unit units) {
        if(units == Unit.kInches)
            return getRangeInches();
        else if(units == Unit.kMillimeters)
            return getRangeMM();
        else
            return -1;
    }

    /**
     * Get the timestamp of the current measurement. Measured in seconds
     * since the program began.
     *
     * Uses Timer.GetFPGATimestamp()
     *
     * @return  timestamp of current measurement
     */
    public double getTimestamp() {
        return m_timestamp;
    }

    /**
     * Enable/disable the sensor in round robin scheduling. Automatic mode
     * must be separately enabled/disabled.
     *
     * @param enable    true - enable device
     *                  false - disable device
     */
    public void setEnabled(boolean enable) {
        m_enabled = enable;
    }

    /**
     * Sets the range profile for the sensor. Valid options are:
     *      kDefault
     *      kHighAccuracy
     *      kLongRange
     *      kHighSpeed
     *
     * Range profiles are based on timing budgets defined in the VL53L0X
     * datasheet
     *
     * If called in automatic mode, the sensor will be stopped until the
     * new profile is set. Measurements will not be available during this
     * period
     *
     * @param profile The range profile to set in the sensor
     *
     * @return Profile successfully changed
     */
    public boolean setRangeProfile(RangeProfile profile) {
        if(profile == m_profile) return true; // ignore the case of no change

        m_newProfile = profile;

        if(m_stopped && !m_automaticEnabled) {
            //System.out.println("Sensor stopped. Changing profile.");
            if(profile == RangeProfile.kHighAccuracy)
                return setProfileHighAccuracy();
            else if(profile == RangeProfile.kLongRange)
                return setProfileLongRange();
            else if(profile == RangeProfile.kHighSpeed)
                return setProfileHighSpeed();
            else
                return setProfileDefault();
        }
        else {
            //System.out.println("Sensor not stopped");
            m_enabled = false;
            return false;
        }
    }

    /**
     * Set the measurement period for the round robin scheduling loop.
     *
     * @param period    Delay in seconds used by the round robin scheduling
     *                  loop. Should be set to a value larger than the
     *                  measurement timing budget configured in the device, or
     *                  measurements will likely return not ready.
     */
    public void setMeasurementPeriod(double period) {
        if(period < 0.01) {
            period = 0.01;
        }
        else if(period > 1) {
            period = 1;
        }

        m_measurementPeriod = period;
    }

    /**
     * Get the measurement period for the round robin scheduling loop
     *
     * @return measurement period in seconds.
     */
    public double getMeasurementPeriod() {
        return m_measurementPeriod;
    }

    /**
     * Set the current DistanceUnit that should be used for the PIDSource base object.
     *
     * @param units The DistanceUnit that should be used.
     */
    public void setDistanceUnits(Unit units) {
        m_units = units;
    }

    /**
     * Get the current DistanceUnit that is used for the PIDSource base object.
     *
     * @return The type of DistanceUnit that is being used.
     */
    public Unit getDistanceUnits() {
        return m_units;
    }

    private synchronized boolean initialize() {
        boolean status = false;
        int statusInt;

        if((statusInt = VL53L0XJNI.ValidateI2C(m_port, m_addr)) != 0) {
            DriverStation.reportError(String.format("Error 0x%08X: Could not communicate" +
                                                    " with Rev 2M sensor over I2C.",
                                                    statusInt), false);
        }

        status = VL53L0XJNI.DataInit(m_port, m_addr);

        if(status)
            status = VL53L0XJNI.GetDeviceInfo(m_port, m_addr);

        if(status)
            status = VL53L0XJNI.StaticInit(m_port, m_addr);

        if(status)
            status = VL53L0XJNI.PerformRefCalibration(m_port, m_addr);

        if(status)
            status = VL53L0XJNI.PerformRefSpadManagement(m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetDeviceMode(1, m_port, m_addr);

        if(status) {
            status = setRangeProfile(RangeProfile.kDefault);
            if(status) {
                m_sensors.add(this);
                m_enabled = true;
            }
        }

        return status;
    }

    private static class DoContinuous extends Thread {
        @Override
        public synchronized void run() {
            boolean allStopped;
            do {
                allStopped = true;
                for (Rev2mDistanceSensor sensor: m_sensors) {
                    if(sensor.isEnabled()) {
                        if(sensor.m_stopped) {
                            if(VL53L0XJNI.StartMeasurement(sensor.m_port, sensor.m_addr)) {
                                sensor.m_stopped = false;
                            }
                        }
                        else {
                            if(VL53L0XJNI.GetMeasurementDataReady(sensor.m_port, sensor.m_addr)) {
                                double currentRange = VL53L0XJNI.GetRangingMeasurementData(sensor.m_port, sensor.m_addr);

                                if(currentRange > 0) {
                                    sensor.m_currentRange = currentRange;
                                    sensor.m_rangeValid = true;
                                    sensor.m_timestamp = Timer.getFPGATimestamp();
                                }
                                else {
                                    sensor.m_rangeValid = false;
                                }
                                VL53L0XJNI.ClearInterruptMask(sensor.m_port, sensor.m_addr);
                            }
                            else {
                                sensor.m_rangeValid = false;
                            }
                        }
                    }
                    else if(!sensor.m_stopped) {
                        if(!sensor.m_stopping) {
                            if(VL53L0XJNI.StopMeasurement(sensor.m_port, sensor.m_addr)) {
                                sensor.m_stopping = true;
                            }
                        }
                        else {
                            if(VL53L0XJNI.GetStopCompletedStatus(sensor.m_port, sensor.m_addr)) {
                                sensor.m_stopped = true;
                                sensor.m_stopping = false;
                                VL53L0XJNI.ClearInterruptMask(sensor.m_port, sensor.m_addr);
                            }
                        }
                    }
                    else if(sensor.m_newProfile != sensor.m_profile) {
                        allStopped = false;

                        if(sensor.m_newProfile == RangeProfile.kHighAccuracy)
                            sensor.m_enabled = sensor.setProfileHighAccuracy();
                        else if(sensor.m_newProfile == RangeProfile.kLongRange)
                            sensor.m_enabled = sensor.setProfileLongRange();
                        else if(sensor.m_newProfile == RangeProfile.kHighSpeed)
                            sensor.m_enabled = sensor.setProfileHighSpeed();
                        else
                            sensor.m_enabled = sensor.setProfileDefault();
                    }

                    if(!sensor.m_stopped) {
                        allStopped = false;
                    }
                }

                Timer.delay(m_measurementPeriod);
            }
            while(!allStopped || m_automaticEnabled);
        }
    }

    private double getRangeInches() {
        if (isRangeValid()) {
            return m_currentRange / 25.4;
        } else {
            return -1;
        }
    }

    private double getRangeMM() {
        if (isRangeValid()) {
            return m_currentRange;
        } else {
            return -1;
        }
    }

    /**
     * Get the range in the current DistanceUnit.
     *
     * @return The range in DistanceUnit
     */
    public double GetRange() {
        switch (m_units) {
        case kInches:
            return getRangeInches();
        case kMillimeters:
            return getRangeMM();
        default:
            return 0.0;
        }
    }

    private boolean setProfileLongRange() {
        //System.out.println("Setting profile to long range");
        boolean status = false;
        status = VL53L0XJNI.SetLimitCheckEnableSigmaFinalRange(1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckEnableSignalRateFinalRange(1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckEnableRangeIgnoreThreshold(0, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckValueSigmaFinalRange(60, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckValueSignalRateFinalRange(0.1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetMeasurementTimingBudgetMicroSeconds(33000, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetVcselPulsePeriodPreRange(18, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetVcselPulsePeriodFinalRange(14, m_port, m_addr);

        if(status)
            m_profile = RangeProfile.kLongRange;

        return status;
    }

    private boolean setProfileHighAccuracy() {
        //System.out.println("Setting profile to high accuracy");
        boolean status = false;
        status = VL53L0XJNI.SetLimitCheckEnableSigmaFinalRange(1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckEnableSignalRateFinalRange(1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckEnableRangeIgnoreThreshold(0, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckValueSigmaFinalRange(18, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckValueSignalRateFinalRange(0.25, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetMeasurementTimingBudgetMicroSeconds(200000, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetVcselPulsePeriodPreRange(14, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetVcselPulsePeriodFinalRange(10, m_port, m_addr);

        if(status)
            m_profile = RangeProfile.kHighAccuracy;

        return status;
    }

    private boolean setProfileHighSpeed() {
        //System.out.println("Setting profile to high speed");
        boolean status = false;
        status = VL53L0XJNI.SetLimitCheckEnableSigmaFinalRange(1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckEnableSignalRateFinalRange(1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckEnableRangeIgnoreThreshold(0, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckValueSigmaFinalRange(32, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckValueSignalRateFinalRange(0.25, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetMeasurementTimingBudgetMicroSeconds(30000, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetVcselPulsePeriodPreRange(14, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetVcselPulsePeriodFinalRange(10, m_port, m_addr);

        if(status)
            m_profile = RangeProfile.kHighSpeed;

        return status;
    }

    private boolean setProfileDefault() {
        //System.out.println("Setting profile to default");
        boolean status = false;
        status = VL53L0XJNI.SetLimitCheckEnableSigmaFinalRange(1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckEnableSignalRateFinalRange(1, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckEnableRangeIgnoreThreshold(0, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckValueSigmaFinalRange(18, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetLimitCheckValueSignalRateFinalRange(0.25, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetMeasurementTimingBudgetMicroSeconds(33823, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetVcselPulsePeriodPreRange(14, m_port, m_addr);

        if(status)
            status = VL53L0XJNI.SetVcselPulsePeriodFinalRange(10, m_port, m_addr);

        if(status)
            m_profile = RangeProfile.kDefault;

        return status;
    }
}